### PR TITLE
Add JSON validation as a CI stage

### PR DESCRIPTION
**Pull request by Alexander-Wilms**
Since I've encountered issues with JSON parsing from time to time, this way you can catch errors early.

Should I collect the errors and print them at the end or is this good enough?

### CODE REVIEW

**Comment by krs0** [comment]
Could this be also enabled for mod repos?

**Comment by Alexander-Wilms** [comment]
Yes, e.g. one could wget the valid_json.py in each repo's CI setup, but setting up CI for each of them might be too much work. Another possibility could be adding another repo with CI which clones all mod repos and then runs the script.

Also, please don't merge yet, maybe I can validate the JSON directly, instead of using the yaml module.

**Comment by Alexander-Wilms** [comment]
Done, now it prints this:

```
python3 .github/validate_json.py 
Validation of CMakePresets.json succeeded
Validation of client/ios/Images.xcassets/Contents.json succeeded
Validation of client/ios/Images.xcassets/AppIcon.appiconset/Contents.json succeeded
Validation of test/testdata/terrainViewMappings.json succeeded
Validation of test/testdata/ObjectPropertyTest/objects.json succeeded
Validation of test/testdata/ObjectPropertyTest/objects.ex.json succeeded
Validation of test/testdata/ObjectPropertyTest/header.json succeeded
Validation of test/testdata/ObjectPropertyTest/surface_terrain.json succeeded
Validation of test/testdata/ObjectPropertyTest/underground_terrain.json succeeded
Validation of test/testdata/rmg/1.json succeeded
Validation of test/testdata/TerrainTest/objects.json succeeded
Validation of test/testdata/TerrainTest/header.json succeeded
Validation of test/testdata/TerrainTest/surface_terrain.json succeeded
Validation of test/testdata/TerrainTest/underground_terrain.json succeeded
Validation of test/testdata/erm/heroVar.json succeeded
Validation of test/testdata/erm/testy.json succeeded
Validation of test/testdata/erm/DO1.json succeeded
Validation of test/testdata/erm/DO2.json succeeded
Validation of test/testdata/erm/re1.json succeeded
Validation of test/testdata/erm/indirectVar.json succeeded
Validation of test/testdata/erm/std.json failed
JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)")
test/testdata/erm/std.json:10:5
Validation of test/testdata/erm/list-manipulation.json failed
JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)")
test/testdata/erm/list-manipulation.json:10:5
Validation of test/testdata/MiniTest/objects.json succeeded
Validation of test/testdata/MiniTest/header.json succeeded
Validation of test/testdata/MiniTest/surface_terrain.json succeeded
Validation of config/mainmenu.json succeeded
Validation of config/ERMU_to_picture.json succeeded
Validation of config/bonuses.json succeeded
Validation of config/resources.json succeeded
Validation of config/randomMap.json succeeded
Validation of config/campaign_regions.json succeeded
Validation of config/terrainViewPatterns.json succeeded
Validation of config/campaignSets.json succeeded
Validation of config/terrains.json succeeded
Validation of config/gameConfig.json succeeded
Validation of config/campaignMedia.json succeeded
Validation of config/heroClasses.json succeeded
Validation of config/battleStartpos.json succeeded
Validation of config/battles_graphics.json succeeded
Validation of config/mapOverrides.json succeeded
Validation of config/obstacles.json succeeded
Validation of config/roads.json succeeded
Validation of config/startres.json succeeded
Validation of config/artifacts.json succeeded
Validation of config/fonts.json succeeded
Validation of config/battlefields.json succeeded
Validation of config/skills.json succeeded
Validation of config/rivers.json succeeded
Validation of config/commanders.json succeeded
Validation of config/filesystem.json succeeded
Validation of config/battleEffects.json succeeded
Validation of config/ambientSounds.json succeeded
Validation of config/spells/timed.json succeeded
Validation of config/spells/adventure.json succeeded
Validation of config/spells/other.json succeeded
Validation of config/spells/moats.json succeeded
Validation of config/spells/offensive.json succeeded
Validation of config/spells/ability.json succeeded
Validation of config/spells/vcmiAbility.json succeeded
Validation of config/schemas/objectTemplate.json succeeded
Validation of config/schemas/script.json succeeded
Validation of config/schemas/skill.json failed
JSONDecodeError("Expecting ',' delimiter: line 78 column 4 (char 2102)")
config/schemas/skill.json:79:5
Validation of config/schemas/heroClass.json succeeded
Validation of config/schemas/terrain.json succeeded
Validation of config/schemas/obstacle.json succeeded
Validation of config/schemas/river.json succeeded
Validation of config/schemas/hero.json succeeded
Validation of config/schemas/settings.json succeeded
Validation of config/schemas/spell.json succeeded
Validation of config/schemas/mapHeader.json succeeded
Validation of config/schemas/creature.json succeeded
Validation of config/schemas/artifact.json succeeded
Validation of config/schemas/template.json succeeded
Validation of config/schemas/objectType.json succeeded
Validation of config/schemas/mod.json succeeded
Validation of config/schemas/object.json succeeded
Validation of config/schemas/faction.json succeeded
Validation of config/schemas/bonus.json succeeded
Validation of config/schemas/battlefield.json succeeded
Validation of config/schemas/townStructure.json succeeded
Validation of config/schemas/townSiege.json succeeded
Validation of config/schemas/road.json succeeded
Validation of config/schemas/townBuilding.json succeeded
Validation of config/creatures/tower.json succeeded
Validation of config/creatures/conflux.json succeeded
Validation of config/creatures/special.json succeeded
Validation of config/creatures/castle.json succeeded
Validation of config/creatures/stronghold.json succeeded
Validation of config/creatures/inferno.json succeeded
Validation of config/creatures/neutral.json succeeded
Validation of config/creatures/rampart.json succeeded
Validation of config/creatures/dungeon.json succeeded
Validation of config/creatures/fortress.json succeeded
Validation of config/creatures/necropolis.json succeeded
Validation of config/factions/tower.json succeeded
Validation of config/factions/conflux.json succeeded
Validation of config/factions/castle.json succeeded
Validation of config/factions/stronghold.json succeeded
Validation of config/factions/random.json succeeded
Validation of config/factions/inferno.json succeeded
Validation of config/factions/neutral.json succeeded
Validation of config/factions/rampart.json succeeded
Validation of config/factions/dungeon.json succeeded
Validation of config/factions/fortress.json succeeded
Validation of config/factions/necropolis.json succeeded
Validation of config/widgets/battleWindow2.json succeeded
Validation of config/widgets/optionsTab.json succeeded
Validation of config/widgets/adventureMap.json succeeded
Validation of config/widgets/turnTimer.json succeeded
Validation of config/widgets/randomMapTab.json succeeded
Validation of config/widgets/settings/library.json succeeded
Validation of config/widgets/settings/settingsMainContainer.json succeeded
Validation of config/widgets/settings/adventureOptionsTab.json succeeded
Validation of config/widgets/settings/battleOptionsTab.json succeeded
Validation of config/widgets/settings/generalOptionsTab.json succeeded
Validation of config/widgets/settings/otherOptionsTab.json succeeded
Validation of config/heroes/tower.json succeeded
Validation of config/heroes/conflux.json succeeded
Validation of config/heroes/special.json succeeded
Validation of config/heroes/castle.json succeeded
Validation of config/heroes/stronghold.json succeeded
Validation of config/heroes/inferno.json succeeded
Validation of config/heroes/rampart.json succeeded
Validation of config/heroes/dungeon.json succeeded
Validation of config/heroes/fortress.json succeeded
Validation of config/heroes/necropolis.json succeeded
Validation of config/objects/generic.json succeeded
Validation of config/objects/creatureBanks.json succeeded
Validation of config/objects/rewardableOnceVisitable.json succeeded
Validation of config/objects/rewardableOncePerHero.json succeeded
Validation of config/objects/rewardablePickable.json succeeded
Validation of config/objects/rewardableOncePerWeek.json succeeded
Validation of config/objects/moddables.json succeeded
Validation of config/objects/dwellings.json succeeded
Validation of config/objects/rewardableBonusing.json succeeded
Validation of .vscode/settings.json succeeded
Validation of Mods/vcmi/mod.json succeeded
Validation of Mods/vcmi/Sprites/PortraitsLarge.json succeeded
Validation of Mods/vcmi/Sprites/PortraitsSmall.json succeeded
Validation of Mods/vcmi/Sprites/ScSelC.json succeeded
Validation of Mods/vcmi/Sprites/itpa.json succeeded
Validation of Mods/vcmi/Sprites/vcmi/battleQueue/statesBig.json succeeded
Validation of Mods/vcmi/Sprites/vcmi/battleQueue/statesSmall.json succeeded
Validation of Mods/vcmi/Sprites/settingsWindow/button32.json succeeded
Validation of Mods/vcmi/Sprites/settingsWindow/button46.json succeeded
Validation of Mods/vcmi/Sprites/settingsWindow/button80.json succeeded
Validation of Mods/vcmi/Sprites/settingsWindow/button190.json succeeded
Validation of Mods/vcmi/Sprites/buttons/commander.json succeeded
Validation of Mods/vcmi/Sprites/buttons/backpack.json succeeded
Validation of Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsGreen.json failed
JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 79)")
Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsGreen.json:6:17
Validation of Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsRed.json failed
JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 77)")
Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsRed.json:6:17
Validation of Mods/vcmi/Sprites/stackWindow/upgradeButton.json succeeded
Validation of Mods/vcmi/Sprites/stackWindow/cancelButton.json succeeded
Validation of Mods/vcmi/Sprites/stackWindow/switchModeIcons.json succeeded
Validation of Mods/vcmi/Sprites/stackWindow/levels.json succeeded
Validation of Mods/vcmi/config/vcmi/spanish.json succeeded
Validation of Mods/vcmi/config/vcmi/french.json succeeded
Validation of Mods/vcmi/config/vcmi/chinese.json succeeded
Validation of Mods/vcmi/config/vcmi/czech.json succeeded
Validation of Mods/vcmi/config/vcmi/towerCreature.json succeeded
Validation of Mods/vcmi/config/vcmi/russian.json succeeded
Validation of Mods/vcmi/config/vcmi/polish.json succeeded
Validation of Mods/vcmi/config/vcmi/german.json succeeded
Validation of Mods/vcmi/config/vcmi/ukrainian.json succeeded
Validation of Mods/vcmi/config/vcmi/english.json succeeded
Validation of Mods/vcmi/config/vcmi/towerFactions.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/heroes3unused/dragon.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/unknownUnused/goldenRing.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/unknownUnused/upgrade.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/unknownUnused/analogy.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/hdmod/jebusCross.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/hdmod/coldshadowsFantasy.json succeeded
Validation of Mods/vcmi/config/vcmi/rmg/hdmod/clashOfDragons.json succeeded
[{'exception': JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)"),
  'position': 'test/testdata/erm/std.json:10:5'},
 {'exception': JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)"),
  'position': 'test/testdata/erm/list-manipulation.json:10:5'},
 {'exception': JSONDecodeError("Expecting ',' delimiter: line 78 column 4 (char 2102)"),
  'position': 'config/schemas/skill.json:79:5'},
 {'exception': JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 79)"),
  'position': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsGreen.json:6:17'},
 {'exception': JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 77)"),
  'position': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsRed.json:6:17'}]
Traceback (most recent call last):
  File "/workspaces/vcmi/.github/validate_json.py", line 31, in <module>
    raise Exception("Not all JSON files are valid")
Exception: Not all JSON files are valid
```

**Comment by krs0** [comment]
```
{'exception': JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)"),
  'position': 'test/testdata/erm/list-manipulation.json:10:5'},
 {'exception': JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)"),
  'position': 'test/testdata/erm/std.json:10:5'}]
```

I cannot figure these 2, any idea?

**Comment by Alexander-Wilms** [comment]
vscode highlights the missing comma:

image

**Comment by krs0** [comment]
I meant these sorry :(

[{'exception': JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 79)"),
  'position': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsGreen.json:6:17'},
 {'exception': JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 77)"),
  'position': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsRed.json:6:17'},

**Comment by Alexander-Wilms** [comment]
I think the leading zeros are the issue:

image

**Comment by Alexander-Wilms** [comment]
The different validation methods:

```
json:  strict, but doesn't preserve line numbers necessarily, since it strips comments before parsing
json5: strict and preserves line numbers even for files will with line comments
yaml:  less strict, allows e.g. leading zeros
```

json:
```
[{'error_msg': JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 79)"),
  'error_pos': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsGreen.json:5:16'},
 {'error_msg': JSONDecodeError("Expecting ',' delimiter: line 5 column 16 (char 77)"),
  'error_pos': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsRed.json:5:16'},
 {'error_msg': JSONDecodeError("Expecting ',' delimiter: line 78 column 4 (char 2102)"),
  'error_pos': 'config/schemas/skill.json:78:4'},
 {'error_msg': JSONDecodeError("Expecting ',' delimiter: line 494 column 6 (char 15346)"),
  'error_pos': 'config/widgets/adventureMap.json:494:6'},
 {'error_msg': JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)"),
  'error_pos': 'test/testdata/erm/list-manipulation.json:9:4'},
 {'error_msg': JSONDecodeError("Expecting ',' delimiter: line 9 column 4 (char 67)"),
  'error_pos': 'test/testdata/erm/std.json:9:4'}]
```

json5:
```
[{'error_msg': ValueError('<string>:5 Unexpected "0" at column 16'),
  'error_pos': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsGreen.json:5:16'},
 {'error_msg': ValueError('<string>:5 Unexpected "0" at column 16'),
  'error_pos': 'Mods/vcmi/Sprites/battle/rangeHighlights/rangeHighlightsRed.json:5:16'},
 {'error_msg': ValueError('<string>:78 Unexpected "]" at column 4'),
  'error_pos': 'config/schemas/skill.json:78:4'},
 {'error_msg': ValueError('<string>:513 Unexpected """ at column 6'),
  'error_pos': 'config/widgets/adventureMap.json:513:6'},
 {'error_msg': ValueError('<string>:9 Unexpected """ at column 4'),
  'error_pos': 'test/testdata/erm/list-manipulation.json:9:4'},
 {'error_msg': ValueError('<string>:9 Unexpected """ at column 4'),
  'error_pos': 'test/testdata/erm/std.json:9:4'}]
```

yaml:
```
[{'error_msg': ParserError('while parsing a flow mapping', <yaml.error.Mark object at 0x7fe5ff467490>, "expected ',' or '}', but got ']'", <yaml.error.Mark object at 0x7fe5ff4659f0>),
  'error_pos': 'config/schemas/skill.json:78:13'},
 {'error_msg': ParserError('while parsing a flow mapping', <yaml.error.Mark object at 0x7fe5ff160cd0>, "expected ',' or '}', but got '<scalar>'", <yaml.error.Mark object at 0x7fe5ff161390>),
  'error_pos': 'config/widgets/adventureMap.json:513:21'}]
```

I think having all 3 methods in the script is helpful, since e.g. the yaml appears to be the closest to the implementation used by VCMI. But only the line and column of the json5 validation should always be correct, even though it's stricter and also complains about numbers with leading zeros.

So for CI, I'd set VALIDATION_TYPE to 'json5'. What do you think?

**Comment by krs0** [comment]
I do not mind deleting the leading zeroes if that is needed. 30s job.

File: .github/validate_json.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import re
```
Comment (kambala-decapitator) [inline .github/validate_json.py:1]: normally scripts should start with shebang

File: .github/validate_json.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+errors = []
+for path in sorted(Path('.').glob('**/*.json')):
+    # because path is an object and not a string
+    path_str = str(path)
+    try:
+        with open(path_str, 'r') as file:
+            if VALIDATION_TYPE == 'json':
+                jstyleson.load(file)
+            if VALIDATION_TYPE == 'json5':
```
Comment (kambala-decapitator) [inline .github/validate_json.py:23]:
```suggestion
            elif VALIDATION_TYPE == 'json5':
```

**Comment by IvanSavenko** [comment]
Thanks!

As for validator selection - json5 looks best option for me too.

File: .github/workflows/github.yml [hunk 3]
```diff
@@ -145,6 +145,13 @@ jobs:
       with:
         submodules: recursive
 
+    - name: Validate JSON
+      # the Python yaml module doesn't seem to work on mac-arm
+      if: ${{ startsWith(matrix.preset, 'linux') }}
```
Comment (IvanSavenko) [inline .github/workflows/github.yml:150]: Running it on multiple presets is probably redundant and slightly increases already long CI built times.
Perhaps run only on `[build (linux-qt6, ubuntu-22.04, 0, linux-clang-test)]`?

This is our fastest Linux-based CI preset

**Comment by Alexander-Wilms** [comment]
The commits should be squashed during the merge

**Review by kambala-decapitator — APPROVED** [review]

**Pull request merged**